include/dali/core/mm/cuda_vm_resource.h

```diff
@@ -331,22 +331,43 @@ class cuda_vm_resource : public memory_resource<memory_kind::device> {
   void va_allocate(size_t min_size) {
     size_t va_size = std::max(next_pow2(min_size), initial_va_size_);
 
-    CUdeviceptr hint = {};
-    va_region *last_va = nullptr;
+    cuvm::CUMemAddressRange va;
+
+    SmallVector<CUdeviceptr, 3> hints;
+
     if (va_regions_.empty()) {
       // Calculate the hint address for the allocations for this device.
       // There's some not very significant base and different devices get
-      // address spaces separated by 2^40 - this should be quite enough.
-      hint = (device_ordinal_ + 1) * 0x10000000000u;
+      // address spaces separated by at least 2^40 - this should be quite enough.
+      hints = {
+        (device_ordinal_ +  1) * (1_u64 << 40),
+        (device_ordinal_ + 17) * (1_u64 << 40),
+        (device_ordinal_ + 33) * (1_u64 << 40),
+      };
     } else {
-      // Try to allocate after the last VA for this device
-      last_va = &va_regions_.back();
-      hint = last_va->address_range.end();
+      // Try to allocate after the last VA for this device or before the first
+      auto &first_va = va_regions_.front();
+      auto &last_va = va_regions_.back();
+      assert(!va_ranges_.empty());
+      va_size = 2 * va_ranges_.back().size();
```

Comment:
Where this formula came from?
Why you need to recalculate `va_size` in this case?

Reply:
Oh, it's a bug - it should be `max`.